Cython/TestUtils.py

```diff
@@ -184,34 +186,35 @@ def visit_CompilerDirectivesNode(self, node):
     visit_Node = VisitorTransform.recurse_to_children
 
 
-def unpack_source_tree(tree_file, dir=None):
-    if dir is None:
-        dir = tempfile.mkdtemp()
-    header = []
-    cur_file = None
-    f = open(tree_file)
-    try:
-        lines = f.readlines()
-    finally:
-        f.close()
-    del f
-    try:
-        for line in lines:
-            if line[:5] == '#####':
-                filename = line.strip().strip('#').strip().replace('/', os.path.sep)
-                path = os.path.join(dir, filename)
-                if not os.path.exists(os.path.dirname(path)):
-                    os.makedirs(os.path.dirname(path))
-                if cur_file is not None:
-                    f, cur_file = cur_file, None
-                    f.close()
-                cur_file = open(path, 'w')
-            elif cur_file is not None:
-                cur_file.write(line)
-            elif line.strip() and not line.lstrip().startswith('#'):
-                if line.strip() not in ('"""', "'''"):
-                    header.append(line)
-    finally:
-        if cur_file is not None:
-            cur_file.close()
-    return dir, ''.join(header)
+def unpack_source_tree(tree_file, workdir, cython_root):
+    programs = {'PYTHON': [sys.executable],
+                'CYTHON': [sys.executable, os.path.join(cython_root, 'cython.py')],
+                'CYTHONIZE': [sys.executable, os.path.join(cython_root, 'cythonize.py')]}
+
+    if workdir is None: workdir = tempfile.mkdtemp()
+    header, cur_file = [], None
+    with open(tree_file) as f:
+        try:
+            for line in f:
+                if line.startswith('#####'):
+                    filename = line.strip().strip('#').strip().replace('/', os.path.sep)
+                    path = os.path.join(workdir, filename)
+                    if not os.path.exists(os.path.dirname(path)):
+                        os.makedirs(os.path.dirname(path))
+                    cur_file.close()
+                    cur_file = open(path, 'w')
+                elif not getattr(cur_file, 'closed', True):
```

Comment:
Why not keep the much simpler `is None`? (Same below) This change makes the unnecessary assumption that whatever `open()` returns has a `closed` attribute that changes with a call to `.close()`. I don't see a need for that assumption.

Reply:
> Why not keep the much simpler `is None`?

You're right, for this case it's only to guard against the case when no output file has ever been opened.

> Same below

In the `finally` clause, we're guarding against two cases:
* No output file has ever been opened: checking for None works fine
* `cur_file` is assigned to a currently opening file by the first `if`: if the statement is reached because an exception is raised, it should try to close the file.  I'm not in favor of a `try` that big, but I'm not capable of making it any more explicit.

> This change makes the unnecessary assumption that whatever `open()` returns has a `closed` attribute that changes with a call to `.close()`.

No assumption is made, [since what ever `open` returns is based on `IOBase`, which define the attribute `closed`](https://docs.python.org/2/library/io.html#io.IOBase.closed).

Reply:
There used to be this code, which maps all cases back to `None`, and which I think is fairly explicit:
```python
               if cur_file is not None:
                    f, cur_file = cur_file, None
                    f.close()
```

Reply:
Although I don't like the way this is handled, I think I'm OK with it if it's safer this way.

Reply:
It's ok to disagree on solutions. Here's why I chose the original code over your proposal.

IMHO (!), `getattr()` has two use cases. The good one is to read an attribute by string name. And the smelly one is to check for an attribute and read it if it's there. It feels smelly because it's fairly broad – there may be many reasons why such an attribute does not exist, and some of them may be bugs or at least unintended. And for the innocent reader, it's difficult to understand what the originally considered and correctly covered reasons were.

In your case, I assume that you were only expecting `None` to hit the "no such attribute" case. That may be so – as long as things really work as expected and no-one replaces the builtin `open`, for example (fine, that's probably acceptable). For a future reader, however, it may take a while to understand that this is the only intended case, and that changing this code will not lead to unintended bugs due to other, not so obvious cases.

If, instead, I write `is not None`, then it's immediately clear that this is the only intended case. And looking at the other usages will somewhat quickly make it clear in which cases `None` can appear. For me, that's a reason to prefer the original version.

Reply:
Thanks for the insight, especially the forward-compatibility/gotchas!